### ISSUE TITLE
Indicate that saving is in progress

### ISSUE THREAD

**Issue by Skeeve** [Jermolene/TiddlyWiki5#1269]
Please indicate somehow that saving is in progress

https://groups.google.com/d/msg/tiddlywiki/l8Z0gdCunuM/SV2R2MgrTfUJ

When saving to a server, you only know that it's done, when you notice the success message.

Maybe we can indicate that saving is in progress by modifying the save button

Easiest enhancement would be a pulsating button as shown below.

But there are more neat tricks available like a pie-timer: http://css-tricks.com/css-pie-timer/

```
<svg class="tc-image-save-button tc-image-button" viewBox="0 0 128 128" width="22pt" height="22pt">
    <g fill-rule="evenodd">
    <path d="M120.78304,34.329058 C125.424287,43.1924006 128.049406,53.2778608 128.049406,63.9764502 C128.049406,99.3226742 99.3956295,127.97645 64.0494055,127.97645 C28.7031816,127.97645 0.0494055385,99.3226742 0.0494055385,63.9764502 C0.0494055385,28.6302262 28.7031816,-0.0235498012 64.0494055,-0.0235498012 C82.8568763,-0.0235498012 99.769563,8.08898558 111.479045,21.0056358 L114.159581,18.3250998 C117.289194,15.1954866 122.356036,15.1939641 125.480231,18.3181584 C128.598068,21.4359957 128.601317,26.5107804 125.473289,29.6388083 L120.78304,34.329058 Z M108.72451,46.3875877 C110.870571,51.8341374 112.049406,57.767628 112.049406,63.9764502 C112.049406,90.4861182 90.5590735,111.97645 64.0494055,111.97645 C37.5397375,111.97645 16.0494055,90.4861182 16.0494055,63.9764502 C16.0494055,37.4667822 37.5397375,15.9764502 64.0494055,15.9764502 C78.438886,15.9764502 91.3495036,22.308215 100.147097,32.3375836 L58.9411255,73.5435552 L41.975581,56.5780107 C38.8486152,53.4510448 33.7746915,53.4551552 30.6568542,56.5729924 C27.5326599,59.6971868 27.5372202,64.7670668 30.6618725,67.8917192 L53.279253,90.5090997 C54.8435723,92.073419 56.8951519,92.8541315 58.9380216,92.8558261 C60.987971,92.8559239 63.0389578,92.0731398 64.6049211,90.5071765 L108.72451,46.3875877 Z"></path>
    </g>
<animate attributeType="CSS" attributeName="opacity"
       values="1; 0; 1"  keytimes="0: 0.5; 1;" dur="2s" repeatCount="indefinite" />
</svg>
```

**Comment by tobibeer** [comment]
> Pulsating is, btw, no issue at all as shown in my example. SVG already supports it.

Animations can look nice but they can also be(come) annoyning... so, I'd be in favor of something simple, which someone who fancies it can extend to a splashy animated something.

**Comment by twMat** [comment]
>All we need is a button of a differnt color, e.g. purple, which only turns grey upon success.

While I'm also not fond of unnecessary blipping, a problem with static colors or similar to indicate loading is that you don't know if it is progressing or if it really got stuck.

I just found [this](http://lab.hakim.se/ladda/) which, among others, has a combined loading indicator and progressbar. (BTW, that guys is doing some crazy CSS stuff. Imagine [this](http://lab.hakim.se/scroll-effects/) when scrolling your sidebar. Or even scrolling the story river!)

**Comment by tobibeer** [comment]
> a problem with static colors or similar to indicate loading is that you don't know if it is progressing or if it really got stuck

As for that problem, there's absolutely no difference between an animated image and something static... both suggest something in progress whereas in reality we're stuck.

**Comment by twMat** [comment]
Hm, the difference is that an indicator bar can behave differently along the progress. If it starts out smooth but halts, thats a pretty good indicator that something is not working.

**Comment by tobibeer** [comment]
>  If it starts out smooth but halts, thats a pretty good indicator that something is not working.

What you are thus suggesting is yet another state that indicates that we have a problem, i.e. in the form of a halted animation or perhaps some icon with an exclamation mark on top or, in fact, red as in "that didn't work".

**Comment by twMat** [comment]
I've kind of been assuming that a progress bar is like e.g a video without any buffering i.e so that it really reflects what is happening. "loaded bytes / total bytes * length of bar" sort of. Not exactly sure how it would know the total bytes tho.

**Comment by twMat** [comment]
Perhaps a "quick and actionable" until something fancier comes up?

I.e, specification for whoever is capable:

1. Change message "Starting to save wiki" into "TiddlyWiki is saving, please wait..." (but, I propose, in color faint red. Yellow would be good if only it wasn't that the alert to signify saving is *finished
 in the non-server version is yellow.)
2. Remove timer or whatever it is that makes this first alert fade/disappear
3. ...and instead use whatever triggers the second message to instead trigger remove of the first
4. ...only to immediately show the second message (but, I propose, in color green).

**Comment by tobibeer** [comment]
As for the colors, I think what you're really after are message types, e.g.

no type — as is
**i** — info — yellow
**!** — attention — red
**✓** — ok — green
**?** — question — maybe yellow, no idea, should probably be a modal instead

So, starting to save triggers the red attention-type-message which is followed by a green ok-type-message.

Personally, I wouldn't use anything scarily red for things like waiting. You shouldn't be indicating to a user that he should be concerned. Instead have an hourglas or something like that, e.g.

... — wait — blue

**Comment by twMat** [comment]
Fair enough, my only concern is that the "waiting" color is not the same color as any (neither server or non-server) "finished" color. Actually, maybe the non-server version "finished" message shoudl be changed to green, so to free up yellow for waiting.

Or just an hourglass as you say, perfectly fine. Propbably even better as it is language independent (=less translations). And the finished being the circled tick but green and which fades in a sec.

**Comment by tobibeer** [comment]
Oh, I wasn't suggesting to only show an hourglass or a tick, but instead a typed message that is colored and features an icon.

**Comment by twMat** [comment]
![alert_saving](https://cloud.githubusercontent.com/assets/4534056/5887121/3578faf8-a3c0-11e4-8e48-d01a2e9f4794.png)

![alert_saved](https://cloud.githubusercontent.com/assets/4534056/5887123/3e23212e-a3c0-11e4-97ec-6cb0b0016c0b.png)

Good? First one staying until saving process finished whereupon it is replaced with second that then fades quickly.

Assuming it's OK; Can we sum it up into this quacktion spec (- why, a *quick and actionable* specification of course!?). Please @Skeeve and @tobibeer rephrase, if you need to, so we can come to a final conclusion:

1. Change current first alert message "Starting to save wiki" into what is seen above in first image
2. Remove timer or whatever it is in current code that makes this first alert fade/disappear
3. ...and instead use whatever triggers the second message to instead trigger removal of the first alert
4. ...only to immediately show the second alert as illustrated above

- and in non server version - 

1. Change current saved confirmation alert to look like the second image here

**Comment by Skeeve** [comment]
Ilke it - but pulsating while savinf would be nice too ;)

**Comment by tobibeer** [comment]
> Ilke it - but pulsating while saving would be nice too ;)

To me it would be one of the things to turn off. I'd rather have some examples somewhere that show people how to make things wiggle and spin, if they want them to.

**Comment by twMat** [comment]
In addition to whatever else may eventually be implemented, I propose also include this little css to get the mouse cursor to show a little pulsating circle.

`cursor:progress; `

This is useful because you don't need to look up in the corner for any indicators.

Note that the cursor should not have to be positioned over some element (such as the save button) for it to show.

**Comment by sukima** [comment]
>I've kind of been assuming that a progress bar is like e.g a video without any buffering i.e so that it really reflects what is happening. "loaded bytes / total bytes * length of bar" sort of. Not exactly sure how it would know the total bytes tho.

@twMat: This is a very complicated issue. Currently the core XmlHTTPRequest implementation in TW doesn't expose the upload progress. And even if it the [`XMLHttpRequestEventTarget`](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequestEventTarget) is a bit spotty and only provides useful feedback on larger files not simple tiddler changes (which would look like 1% -> 99%) not the slow progression (1%…10%…20%…etc.) you would want. And that is only *half* the story since the whole progress concept is for uploads only. The download part is not included.

In any event the realistic progress you're looking for would be haphazard at best. In most all use cases it would be more a binary (starting … finished) set of events which would not suit well to a progress bar.

A better user experience is to *fake* it. And this is pretty much what the industry does in general. Even Apple's [boot progress bar is fake](http://daringfireball.net/misc/2005/04/tiger_details#waitingforloginwindow) (and also why Windows booting shows an infinite animation and no concept of estimating completion).

![Windows XP Boot Progress Bar](http://1.bp.blogspot.com/-PmMIYZPPmko/UimzIeKO_3I/AAAAAAAAAkk/q479uC_OSzI/s1600/1.1.gif)

The truth of the progress bar can be summed up in this article: "[The Placebo Mini-Pattern][1]".

I would instead recommend that the notification content stick to messages while a small (say 5px) high bar appears below the side toolbar which essentially *fakes* progress using a timer in the same manor described by the [Placebo Mini-Patter][1].

[1]: http://fishbowl.pastiche.org/2003/08/01/the_placebo_minipattern/

**Comment by twMat** [comment]
@sukima - thanks for informative and fun info.

I think we can "temporarily conclude" that 95% of the need is to just get **any** indicator to display between clicking save/upload and finished. The remaining 5% is just pretty. [Numbers reached via diff-X5 analysis].
